klippy/extras/output_pin.py

```diff
@@ -66,5 +81,15 @@ def cmd_SET_PIN(self, gcmd):
         toolhead.register_lookahead_callback(
             lambda print_time: self._set_pin(print_time, value, cycle_time))
 
+    def _resend_current_val(self, eventtime):
+        toolhead = self.printer.lookup_object('toolhead')
+        toolhead.register_lookahead_callback(lambda print_time:
+            self._set_pin(self.last_print_time + 0.8 * self.safety_timeout,
+                           self.last_value, self.last_cycle_time, True)
+            )
```

Comment:
Thanks.  Unfortunately, it's not valid to call `toolhead.register_lookahead_callback()` without holding the gcode mutex.  That is, it can't be called directly from a reactor callback.  You probably want to do something like `print_time = self.mcu_pwm.get_mcu().estimated_print_time(eventtime)` instead - though the time will need to be padded by 100ms (to allow for transmission time to the mcu).

-Kevin

Reply:
Is it allowed to just set the pin directly, as it is now?
100ms time padding is now explicitly in the "relax_margin" which is supposed to give the scheduler some time.

Reply:
The `toolhead.get_last_move_time()` and related `toolhead.register_lookahead_callback()` always return a time in the future, so that it is possible to schedule events at that time.  A call to `mcu.estimated_print_time(systime)` returns the time of "now" on the micro-controller - if one were to schedule an event at that time, it would always result in a "timer too close" as it takes time to transmit that schedule to the micro-controller.

I'm not sure what the goal of the "relax_margin" is.  You only need the PIN_MIN_TIME - any additional time is probably not a good thing as it limits how response future updates are.

-Kevin

Reply:
the "relax margin" was supposed to reduce stress between "now" and the scheduled time the pin has to update (and to reduce the possibility of too close timers).
But if you say that this just removes responsiveness, it can be replaced for just the PIN_MIN_TIME, instead of 10% timeout + PIN_MIN_TIME.